plutus-metatheory/src/Type/RenamingSubstitution.lagda.md

```diff
@@ -25,20 +25,35 @@ open _⊢⋆_
 
 ## Type renaming
 
-A type renaming is a mapping of type variables to type variables.
+Types can contain functions and as such are subject to a nontrivial equality relation. To
+explain the computation equation (the β-rule) we need to define substitution for a single
+type variable in a type. A type renaming is a function from type variables in one context to
+type variables in another. We only need the ability to
+introduce new variable on the right hand side of the context. The simplicity of the
+definition makes it easy to work with and we get some properties for free that we would
+have to pay for with a first order representation, such as not needing to define a lookup
+function, and we inherit the properties of functions provided by η-equality, such as
+associativity of composition, for free. Note that even though
+renamings are functions we do not require our metatheory (Agda’s type system) to support
+functional extensionality. We only ever need to make use of an equation
+between renamings on a point (a variable) and therefore need only a pointwise version
+of equality on functions to work with equality of renamings and substitutions.
 
 ```
 Ren : Ctx⋆ → Ctx⋆ → Set
 Ren Φ Ψ = ∀ {J} → Φ ∋⋆ J → Ψ ∋⋆ J
```

Comment:
I learned the beauty of [order-preserving embeddings](https://fplab.bitbucket.io/posts/2008-03-07-order-preserving-embeddin.html) from James, I feel betrayed.

Reply:
Ha! We're doing stuff in a higher order way here by using NBE. Renaming and substitution as functions are quite nice to work with unless you run into issues with extensionality which we don't here. However you're right to notice that we don't need to full power of renamings but that doesn't bite us either. There is also a nice correspondence with relative monads which gives us a clear framework to work out what properties we need.

Reply:
Also amazed by the link!